### PR TITLE
chore: optimise older opcodes in reverse order

### PR DESCRIPTION
**Pull request by guipublic**
# Description

## Problem\*

Related to #6451 and #6461

## Summary\*
Handle the case where we can optimise an older opcode

## Documentation\*

Check one:
- [X] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Comment by guipublic** [comment]
The increase in circuit size is not expected. This is worrying and we should not merge the PR which is supposed to improve (only a little bit but still) this number.

**Comment by TomAFrench** [comment]
Closes https://github.com/noir-lang/noir/issues/6668?

File: acvm-repo/acvm/src/compiler/optimizers/merge_expressions.rs
Comment (aakoshh) [inline acvm-repo/acvm/src/compiler/optimizers/merge_expressions.rs:98]: Huh, if I knew these existed I might have tried to use them in https://github.com/noir-lang/noir/pull/6695
Although there I was trying to avoid having to allocate new sets along the way.

File: acvm-repo/acvm/src/compiler/optimizers/merge_expressions.rs
Comment (aakoshh) [inline acvm-repo/acvm/src/compiler/optimizers/merge_expressions.rs:145]: This could also be rewritten in terms of [WitnessFolder::fold_brillig_inputs](https://github.com/noir-lang/noir/blob/70cb16a2d38a1d6dc893371ade88e453ccc99070/acvm-repo/acvm/src/compiler/transformers/mod.rs#L324)

File: acvm-repo/acvm/src/compiler/optimizers/merge_expressions.rs
Comment (aakoshh) [inline acvm-repo/acvm/src/compiler/optimizers/merge_expressions.rs:105]: Does this need to be cloned and re-inserted?  Maybe this would work:
```suggestion
                                            used_witness.entry(w2).and_modify(|v| {.clone();
                                                v.insert(b);
                                                v.remove(&i);
                                            };
```

File: acvm-repo/acvm/src/compiler/optimizers/merge_expressions.rs
Comment (aakoshh) [inline acvm-repo/acvm/src/compiler/optimizers/merge_expressions.rs:121]:
```rust
// Merge the lower into the higher opcode.
if i != b && merge_opcodes(i.max(b), i.min(b)) {
    // We need to stop here and continue with the next opcode
    // because the merge invalidates the current opcode.
    break;
}
```
Just an idea, feel free to ignore.

**Review by aakoshh — APPROVED** [review]
This looks good as far as I understand it. 

So to be clear, what's happening is that previously we went through opcodes 0, 1, 2, etc and only considered merging `i` into `b` if `i < b`, so earlier opcodes got merged into later opcodes. We did this when we found a witness `w` that was only used by `i` and `b`. 

However, there could have been a witness `w'` that was used by `i`, `j` and `k`, and after merging `i` into `b` we may or may not have recognised that we can merge `j` into `k` now that `w'` was only used in two opcodes: if we visited `j` earlier than `i` then we wouldn't eliminate it. 

But now we can, because when we hit `k` we see that `j` can be merged into it 🎉 

Does this completely eliminate the need for doing multiple passes with the expression merger?

**Comment by aakoshh** [comment]
> Does this completely eliminate the need for doing multiple passes with the expression merger?

I guess it doesn't, because if for example we have a witness that is only used by the last two opcodes, and we eliminate the one before last, it might unlock the merger of earlier opcodes that we won't be visiting again. 

For example:
Opcode | Witness A | Witness B 
:---: | :---: | :---:
0 | x | 
1 | x | 
2 | x | x
3 |   | x

When we reach opcode 2 we recognise that B is only used by 2 and 3, so we merge 2 into 3, but it's too late to merge 0 into 1 now that they are the only ones using A.

File: acvm-repo/acvm/src/compiler/optimizers/merge_expressions.rs
Comment (aakoshh) [inline acvm-repo/acvm/src/compiler/optimizers/merge_expressions.rs:104]:
```suggestion
                                            v.insert(target);
                                            v.remove(&source);
```
`i` and `b` could be in any order now, right? We could be removing either one.

File: acvm-repo/acvm/src/compiler/optimizers/merge_expressions.rs
Comment (guipublic) [reply acvm-repo/acvm/src/compiler/optimizers/merge_expressions.rs:105]: Good point!

File: acvm-repo/acvm/src/compiler/optimizers/merge_expressions.rs
Comment (guipublic) [reply acvm-repo/acvm/src/compiler/optimizers/merge_expressions.rs:104]: Nice spot!

File: acvm-repo/acvm/src/compiler/optimizers/merge_expressions.rs
Comment (guipublic) [reply acvm-repo/acvm/src/compiler/optimizers/merge_expressions.rs]: Since merge_opcodes can now handled the two cases, I put this logic inside the function, I think it is a bit nicer like this.

**Comment by guipublic** [comment]
> > Does this completely eliminate the need for doing multiple passes with the expression merger?
> 
> I guess it doesn't, because (...)

Yes I agree it does not eliminate all in one pass.

**Review by aakoshh — APPROVED** [review]
Nice!

**Pull request merged**
